compiler/noirc_frontend/src/parser/parser/types.rs

```diff
@@ -201,8 +203,16 @@ pub(super) fn generic_type_args<'a>(
         // separator afterward. Failing early here ensures we try the `type_expression`
         // parser afterward.
         .then_ignore(one_of([Token::Comma, Token::Greater]).rewind())
-        .or(type_expression()
-            .map_with_span(|expr, span| UnresolvedTypeData::Expression(expr).with_span(span)))
+        .or(numeric_generic_type_expression().validate(|(type_arg, err_expr), span, emit| {
+            if let UnresolvedTypeData::Error = type_arg.typ {
+                let err_expr = err_expr.expect("Error expression is expected with an error type");
```

Comment:
```suggestion
            if let Some(err_expr) = err_expr {
```
I'm also not really a fan of the 2 step verification of map_with_span in `numeric_generic_type_expression` and the `validate` here. It'd be simpler if we just did it all here I think - it also gets rid of the "returning a tuple instead of an enum on error" pattern.